[BOLT][AArch64] Provide createDummyReturnFunction

paschalis-mpeis opened the pull request:
On AArch64, this method is needed when trying to instrument a static
binary.

Sample commands:
```bash
clang -Wl,-q test.c -static -o out
llvm-bolt -instrument -instrumentation-sleep-time=5 out -o out.instr
```

---





### Unrelated runtime crash with static binaries
Currently, on AArch64 there is an unrelated runtime crash (at binary loading time to be specific) with any rewritten > static binaries. Trying to invoke any of such binaries (including ones with instrumentation) will result to:
> Unexpected reloc type in static binary.

This will be a separate patch:
- #97710

paschalis-mpeis committed:
[BOLT][AArch64] Instrumentation for static binaries doesn't work in reloc mode.

On AArch64, when trying to instrument a static binary that has relocation
data BOLT would crash as it misses `createDummyReturnFunction` function.

paschalis-mpeis committed:
[BOLT][AArch64] Implemented createDummyReturnFunction.

On AArch64, this method is needed when trying to instrument a static
binary.

Sample commands:
```bash
clang -Wl,-q test.c -static -o out
llvm-bolt -instrument -instrumentation-sleep-time=5 out -o out.instr
```

samolisov commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -1582,6 +1607,12 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
+  InstructionListType createDummyReturnFunction(MCContext *Ctx) const override {
+    InstructionListType Insts(1);
+    createReturn(Insts[0]);
+    return Insts;
```
The code is exactly the same as in `X86MCPlusBuilder.cpp`. What about to uplift the code into the implementation of the `MCPlusBuiler` base class?

paschalis-mpeis replied:
True, and initially I was tempted to do this. But then I thought it might be better for other targets to be explicitly aware of this workaround that gets used by Instrumentation.

Another argument, is that it might let new targets omit such a test. This could have been true for AArch64 if it was handled already like that. To be a better example, I've improved the AArch64 in my latest patch.

These are my thoughts, and not an absolute opinion. What do you think?

samolisov replied:
I believe, a common implementation of a method in a basic class is a good thing until we mark the method as a pure virtual function (what is not our case). When a new architecture needs a different implementation of the method, it should just override it. I agree, eventually this may lead to a situation where two architectures share the common implementation while many many other do not. A controversial thing.

I agree, a lot of method in the base class here have only some stub implementation with an error reporting and `return false`, this is an approach used trough the whole llvm codebase. It strange a little bit, but I get the goal: just let the developers of particular architectures do not implement all of them.

paschalis-mpeis replied:
Yeap, I also agree and see the pros/cons of both sides.

Just to be clear, your preference is still a base class implementation or making this method pure virtual? (but the rest of the MCBuilder will remain as is).

samolisov replied:
My opinion is to put the code in the base class' method. I'm not sure using pure virtual methods is consistent with the rest of bolt's codebase.

paschalis-mpeis replied:
Thank you. Given how simple the implementation is and that AArch64 has testing, I have no objections to this. Implementation is now uplifted to the base class.

paschalis-mpeis commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -706,8 +706,20 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
+      //   adr     x6, 0x219fb0 <sigall_set+0x88>
+      //   add     x6, x6, x14, lsl #2
+      //   ldr     w7, [x6]
+      //   add     x6, x6, w7, sxtw => no shift amount
```
Hey @samolisov ,

This particular change, and a few others, are **not related** to my patch. :)

I have modified only the files: `dummy-return.test`, `main.c`, and `AArch64MCPlusBuilder.cpp`.
But you are seeing more changes because this is a 'stacked pull-request' meaning, that it depends on:
- https://github.com/llvm/llvm-project/pull/83394

Therefore, it also shows all changes of the above Pull-Request as well.

So, to see only the relevant changes, you must click on individual commits, like [this one](https://github.com/llvm/llvm-project/pull/96626/commits/7719e07d74a98f580836d3e32f6133e994be4f98) from the ['Commits' tab](https://github.com/llvm/llvm-project/pull/96626/commits).
I document this on the section 'NOTES' on my initial comment [here](https://github.com/llvm/llvm-project/pull/96626#issue-2372592692).

PS, even if this gets accepted, it will be merged only after #83394 was merged.

paschalis-mpeis committed:
Refactored, improved test, and addressed reviewers.

paschalis-mpeis commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -1582,6 +1607,12 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
     return Instrs;
   }
 
+  InstructionListType createDummyReturnFunction(MCContext *Ctx) const override {
```
True, and initially I was tempted to do this. But then I thought it might be better for other targets to be explicitly aware of this workaround that gets used by Instrumentation.

Another argument, is that it might let new targets omit such a test. This could have been true for AArch64 if it was handled already like that. To be a better example, I've improved the AArch64 in my latest patch.

These are my thoughts, and not an absolute opinion. What do you think?

paschalis-mpeis replied:
Done, thanks.

paschalis-mpeis commented:
Hi  samolisov,
Thanks for your review; I hope I've addressed your concerns.

Since you requested some minor generic refactoring, I renamed the function to `createDummyReturn`, as we are simply populating the function body in this case. Let me know if you think it needs more adjustments.

Also, I've added you as a reviewer to #83394 a few days back (the parent PR this patch is stacked upon), in case you wanted to have a look. In either case, it looks like there is a consensus and that patch is going forward soon.

yota9 commented:
Hi please rebase the PR

paschalis-mpeis commented:
Now that #83394 is merged, I was able to rebase to main.

@yota9, can you please have a look now, sharing your thoughts on @samolisov concerns [here](https://github.com/llvm/llvm-project/pull/96626#discussion_r1659812304) as well?


Thanks in advance,
Paschalis

paschalis-mpeis committed:
Addressing reviewers (1): Uplifted 'createDummyReturn' to base class

samolisov reviewed (approved):
Thanks for addressing, LGTM.

yota9 reviewed (approved):
Thanks!

aaupov commented on bolt/include/bolt/Core/MCPlusBuilder.h:
```diff
@@ -2041,10 +2041,10 @@ class MCPlusBuilder {
     return InstructionListType();
   }
 
-  virtual InstructionListType createDummyReturnFunction(MCContext *Ctx) const {
```
Basically you're moving the definition of `createDummyReturnFunction` from X86 to MCPlusBuilder, right? Please avoid renaming the method (I don't see what's achieved by it), and move the definition here so it's clear why we don't stub it with llvm_unreachable

yota9 replied:
@aaupov  I'm ok with both namings, but just to be clear this function doesn't really create "function" it only create instructions to make return, creation of function it self is outside of this method, so maybe there is logic to rename it to "createDummyReturn". Or even "createReturn", what is dummy about that IDK :)

aaupov replied:
Agree but let's keep renaming separate. Can rename as follow-up NFC change

paschalis-mpeis replied:
Thanks for your review. Done.

@yota9 yes, that was my intuition; I found implying that a function was created somewhat misleading.
`createReturn` is already used though, so we might go with something like: `createReturnBody`.
I will put up a follow-up NFC patch soon.

aaupov reviewed (changes requested):
Thanks for working on this. Left a comment inline. Please also retitle as an imperative statement: "[BOLT][AArch64] Provide createDummyReturnFunction".

paschalis-mpeis committed:
Addressing reviewers (2)

aaupov commented on bolt/test/AArch64/dummy-return.test:
```diff
@@ -0,0 +1,13 @@
+
+REQUIRES: system-linux
+
+RUN: %clang %p/../Inputs/main.c -o %t -Wl,-q -static
```
```suggestion
RUN: %clang %cflags %p/../Inputs/main.c -o %t -Wl,-q -static
```

paschalis-mpeis replied:
Thanks. I applied the suggestion, however, this would break my current c test as it interferes with `-static`.
I will be converting the test to assembly with a follow-up patch.

paschalis-mpeis committed:
Update bolt/test/AArch64/dummy-return.test

Co-authored-by: Amir Ayupov <aaupov@fb.com>

paschalis-mpeis committed:
Converted test to assembly for '%cflags' compatibility.

The '-static' option does not work well with '-nostartfiles -nostdlib'
that comes with '%cflags', but the latter is needed for better buildbot
compatibility.

I converted the test to assembly and dropped the previous c test.

Dropped dummy-return.test and main.c Input file.

paschalis-mpeis commented:
### Force-pushed history:

Adding `%cflags` interfered with my previous c test `dummy-return.test`. I preferred not to expand that test (ie adding a `_start` method and letting the linker know about the entry method), but instead I've written a minimal assembly test.

I force-pushed, as I wanted to introduce the asm test on the initial 'showcase-commit'.

paschalis-mpeis committed:
Ignore test on X86
